nixos/modules/services/taler/exchange.nix

```diff
@@ -0,0 +1,242 @@
+      description = "Slice for GNU taler exchange processes";
+      before = [ "slices.target" ];
+    };
+
+    systemd.services =
+      lib.genAttrs (map (n: "taler-exchange-${n}") services) (name: {
+        serviceConfig = {
+          DynamicUser = true;
+          User = name;
+          Group = groupName;
+          ExecStart =
+            "${lib.getExe' this.package name} -c ${config.services.taler.configFile}"
+            + lib.optionalString this.debug " -L debug";
+          RuntimeDirectory = name;
+          StateDirectory = name;
+          CacheDirectory = name;
+          ReadWritePaths = [ runtimeDir ];
+          # TODO more hardening
+          # PrivateTmp = "yes";
+          # PrivateDevices = "yes";
+          # ProtectSystem = "full";
+          # Slice = "taler-exchange.slice";
+        };
+        requires = [ "taler-exchange-dbinit.service" ];
+        after = [ "taler-exchange-dbinit.service" ];
+        wantedBy = [ "multi-user.target" ]; # TODO slice?
+      })
+      // {
+        taler-exchange-dbinit = {
+          path = [ config.services.postgresql.package ];
+          script =
+            let
+              # Taken from https://docs.taler.net/taler-exchange-manual.html#exchange-database-setup
+              # TODO Why does aggregator need DELETE?
+              dbScript = pkgs.writers.writeText "taler-exchange-db-permissions.sql" ''
+                GRANT SELECT,INSERT,UPDATE,DELETE ON ALL TABLES IN SCHEMA exchange TO "taler-exchange-aggregator";
+                GRANT SELECT,INSERT,UPDATE ON ALL TABLES IN SCHEMA exchange TO "taler-exchange-closer";
+                GRANT SELECT,INSERT,UPDATE ON ALL TABLES IN SCHEMA exchange TO "taler-exchange-wirewatch";
+                GRANT USAGE ON SCHEMA exchange TO "taler-exchange-aggregator";
+                GRANT USAGE ON SCHEMA exchange TO "taler-exchange-closer";
+                GRANT USAGE ON SCHEMA exchange TO "taler-exchange-wirewatch";
+              '';
+            in
+            ''
+              ${lib.getExe' this.package "taler-exchange-dbinit"}
+
+              psql -f ${dbScript}
+            '';
+          requires = [ "postgresql.service" ];
+          after = [ "postgresql.service" ];
+          serviceConfig = {
+            Type = "oneshot";
+            DynamicUser = true;
+            User = dbName;
+          };
+        };
+      }
+      // {
+        taler-exchange-accounts = {
+          script = lib.concatStringsSep "\n" (
+            map (
+              account: "${lib.getExe' this.package "taler-exchange-offline"} upload < ${account}"
+            ) this.enableAccounts
+          );
+          requires = [ "taler-exchange-httpd.service" ];
+          after = [ "taler-exchange-httpd.service" ];
+          serviceConfig = {
+            Type = "oneshot";
+            DynamicUser = true;
+            User = "taler-exchange-accounts";
+          };
+        };
+      };
+
+    users.groups.${groupName} = { };
```

Comment:
This might be unnecessary since you use `DynamicUser`. From `man systemd.exec`:

> When used in conjunction with DynamicUser= the user/group name specified is dynamically allocated at the time the service is started, and released at the time the service is stopped — unless it is already allocated statically (see below). If DynamicUser= is not used the specified user and group must have been created statically in the user database no later than the moment the service is started, for example using the sysusers.d(5) facility, which is applied at boot or package install time. If the user does exist by then program invocation will fail.

Or do you need the group before the service is started/after it's stopped?